src/llms/llm_loader.py

```diff
@@ -0,0 +1,233 @@
+# workaround to disable UserWarning
+import warnings
+warnings.simplefilter("ignore", UserWarning)
+
+from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
+from langchain.callbacks.manager import CallbackManager
+
+import os, inspect
+
+from utils.logger import Logger
+
+class LLMLoader:
+    """
+    Note: This class loads the LLM backend libraries if the specific LLM is loaded.
+    Known caveats: Currently supports a single instance/model per backend
+
+    The llm_backends parameter is an array with the backend llm and (optional) parameters to override
+
+    # format for dictionary
+    llm_backend = {
+        'backend': 'openai',
+        'params': {...}     # optional key
+    }
+
+    # overriding specific parameters
+    llm_backend = {
+        'backend' : 'ollama',
+        'params': {'temperature': 0.02, 'top_p': 0.95},
+    }
+
+    # using default parameters
+    llm_backend = {'backend': 'tgi'}
+
+    # Using the class
+    llm_config = LLMLoader(llm_backend=llm_backend)
+    llm_chain = LLMChain(llm=llm_config.llm, prompt=prompt)
+
+    """
+    def __init__(self, llm_backend={}, logger=None) -> None:
+        self.logger = logger if logger is not None else Logger("llm_loader").logger
+        # LLM backend to use or fallback to default one
+        self.llm_backend = llm_backend.get('backend') if llm_backend.get('backend') is not None else os.environ.get('LLM_DEFAULT', {'backend':'ollama'})
+        self.llm_params = llm_backend.setdefault('params',{}) # return empty dictionary if not present
+        self.llm = None
+        self._set_llm_instance()
+
+    def _set_llm_instance(self):
```

Comment:
```suggestion
    def _set_llm_instance(self):
      self.logger.debug(f"[{inspect.stack()[0][3]}] Loading LLM instances with default {self.llm_backend}")
  
      llm_instances = {
          'openai': self._openai_llm_instance,
          'ollama': self._ollama_llm_instance,
          'tgi': self._tgi_llm_instance,
          'watson': self._watson_llm_instance,
          'bam': self._bam_llm_instance,
      }
  
      llm_instance_method = llm_instances.get(self.llm_backend)
      if llm_instance_method:
          llm_instance_method()
      else:
          self.logger.error(f"ERROR: Unsupported LLM {self.llm_backend}")
```
Can we just rewrite this logic like this? I am not much inclined towards growing ladders.

Reply:
> Can we just rewrite this logic like this? I am not much inclined towards growing ladders.

this seems like a bit of a personal preference.  Is there more motivation behind why your approach is preferred?  If there are real maintainability, complexity, or bug/side-effect risks that's a different situation.

Reply:
just my .02 cents on readability. Also if someone wants to add/remove an instance in future, they can just edit the`llm_instances` map

Reply:
+1 for mapping instead switches

Reply:
This is a matter of preference. While I understand the request and the increase in memory utilization by using the dictionary will be about ~1520 Bytes (with those 5 and will continue increasing as entries are added) versus a constant of ~192 Bytes when using the `match` statement, it also introduces a limitation in which, if an LLM instance requires additional parameters during invocation, it will go back to requiring `if-elseif-elseif` or going back to `match`.   Because of these reasons, I want to avoid arrays to map functions at this time.